### PR TITLE
Prevent registering an instance of a listener more than once

### PR DESCRIPTION
**Pull request by kashike**
This prevents a single instance of a listener from being registered many times.

https://github.com/SpongePowered/SpongeCommon/issues/394
https://github.com/SpongePowered/SpongeForge/issues/295
https://github.com/SpongePowered/SpongeForge/issues/433

### CODE REVIEW

File: src/main/java/org/spongepowered/common/event/SpongeEventManager.java [hunk 1]
```diff
@@ -140,6 +141,12 @@ public void registerListener(PluginContainer plugin, Object listenerObject) {
         checkNotNull(plugin, "plugin");
         checkNotNull(listenerObject, "listener");
 
+        if (this.registeredListeners.contains(listenerObject)) {
+            SpongeImpl.getLogger().warn("Plugin {} attempted to register an already registered listener ({})", plugin.getId(),
+                    listenerObject.getClass().getName());
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/common/event/SpongeEventManager.java]: Would be nice to dump the stack trace as well. I have a private Bukkit plugin which (I believe) is registering a certain listener twice, and I don't know where the second registration is.

Comment (kashike) [inline src/main/java/org/spongepowered/common/event/SpongeEventManager.java]: alright

File: src/main/java/org/spongepowered/common/event/SpongeEventManager.java [hunk 2]
```diff
@@ -68,6 +68,7 @@
     private final AnnotatedEventListener.Factory handlerFactory = new ClassEventListenerFactory("org.spongepowered.common.event.listener",
             new FilterFactory("org.spongepowered.common.event.filters", classLoader), classLoader);
     private final Multimap<Class<?>, RegisteredListener<?>> handlersByEvent = HashMultimap.create();
+    private final List<Object> registeredListeners = Lists.newArrayList();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/event/SpongeEventManager.java]: Can you make this a `Set` instead?

Comment (kashike) [inline src/main/java/org/spongepowered/common/event/SpongeEventManager.java]: done

**Pull request merged**